VirtoCommerce.Platform.Modules/LoadContextAssemblyResolver.cs

```diff
@@ -1,122 +1,191 @@
+        }
+
+        /// <summary>
+        /// Loads assembly to AssemblyLoadContext.Default or gets it from own assembly cache.
+        /// <para>
+        /// Note that only one version of assembly would be loaded and cached by AssemblyName.Name, for all other versions returns cached assembly.
+        /// </para>
+        /// </summary>
+        /// <param name="managedLibrary">ManagedLibrary object containing library name and paths.</param>
+        /// <returns>Retures loaded assembly (could be cached).</returns>
+        private Assembly LoadAssemblyCached(ManagedLibrary managedLibrary)
+        {
+            var assemblyName = managedLibrary.Name;
+            if (_loadedAssemblies.ContainsKey(assemblyName.Name))
+            {
+                return _loadedAssemblies[assemblyName.Name];
             }
-            return assembly;
+
+            var loadedAssembly = LoadAssemblyInternal(managedLibrary);
+            if (loadedAssembly != null)
+            {
+                _loadedAssemblies.Add(assemblyName.Name, loadedAssembly);
+            }
+            return loadedAssembly;
+        }
+
+        /// <summary>
+        /// Performs loading into AssemblyLoadContext.Default using LoadFromAssemblyName for TPA assemblies and LoadFromAssemblyPath for other dependecies.
+        /// <para>
+        /// Based on https://github.com/natemcmaster/DotNetCorePlugins/blob/8f5c28fa70f0869a1af2e2904536268f184e71de/src/Plugins/Loader/ManagedLoadContext.cs Load method,
+        /// but avoided FileNotFoundException from LoadFromAssemblyName trying only load TPA assemblies that way.
+        /// </para>
+        /// </summary>
+        /// <param name="managedLibrary">ManagedLibrary object containing assembly name and paths.</param>
+        /// <returns>Returns loaded assembly.</returns>
+        private Assembly LoadAssemblyInternal(ManagedLibrary managedLibrary)
+        {
+            // To avoid FileNotFoundException for assemblies that are included in TPA - we load them using AssemblyLoadContext.Default.LoadFromAssemblyName.
+            var assemblyFileName = Path.GetFileName(managedLibrary.AppLocalPath);
+            if (TPA.ContainsAssembly(assemblyFileName))
+            {
+                var defaultAssembly = AssemblyLoadContext.Default.LoadFromAssemblyName(managedLibrary.Name);
+                if (defaultAssembly != null)
+                {
+                    return defaultAssembly;
+                }
+            }
+
+            if (SearchForLibrary(managedLibrary, out var path))
+            {
+                return AssemblyLoadContext.Default.LoadFromAssemblyPath(path);
+            }
+
+            return null;
+        }
+
+        private bool SearchForLibrary(ManagedLibrary library, out string path)
+        {
+            // 1. Check for in _basePath + app local path
+            var localFile = Path.Combine(_basePath, library.AppLocalPath);
```

Comment:
this code cognitive  complexity can be  simplified by generation array of search paths and do single linq call `FirstOrDefault`

Reply:
Candidates pregeneration will add significant overhead. The common scenario would be just finding dependency inside _basePath, which generally speaking means one string creation, comparison and return. Sometimes dependency could be found inside _additionalProbingPaths, it could add couple of string creations. If we will pregenerate all the strings, method would become about 5 times slower.
And i think algorithm is readable and self-described, probably algorithm readability would not change much if we switch to pregeneration array.
Is generalization here costs 5-times performance loss?